src/test/basics/Number_test.cpp

```diff
@@ -0,0 +1,607 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2022 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/basics/IOUAmount.h>
+#include <ripple/basics/Number.h>
+#include <ripple/beast/unit_test.h>
+#include <ripple/protocol/STAmount.h>
+#include <sstream>
+#include <tuple>
+
+namespace ripple {
+
+class saveNumberRoundMode
+{
+    Number::rounding_mode mode_;
+
+public:
+    ~saveNumberRoundMode()
+    {
+        Number::setround(mode_);
+    }
+    explicit saveNumberRoundMode(Number::rounding_mode mode) noexcept
+        : mode_{mode}
+    {
+    }
+    saveNumberRoundMode(saveNumberRoundMode const&) = delete;
+    saveNumberRoundMode&
+    operator=(saveNumberRoundMode const&) = delete;
+};
+
+class Number_test : public beast::unit_test::suite
+{
+public:
+    void
+    testZero()
+    {
+        testcase("zero");
+
+        Number const z{0, 0};
+
+        BEAST_EXPECT(z.mantissa() == 0);
+        BEAST_EXPECT(z.exponent() == Number{}.exponent());
+
+        BEAST_EXPECT((z + z) == z);
+        BEAST_EXPECT((z - z) == z);
+        BEAST_EXPECT(z == -z);
+    }
+
+    void
+    test_limits()
+    {
+        testcase("test_limits");
+        bool caught = false;
+        try
+        {
+            Number x{10'000'000'000'000'000, 32768};
+        }
+        catch (std::overflow_error const&)
+        {
+            caught = true;
+        }
+        BEAST_EXPECT(caught);
+        Number x{10'000'000'000'000'000, 32767};
+        BEAST_EXPECT((x == Number{1'000'000'000'000'000, 32768}));
+        Number z{1'000'000'000'000'000, -32769};
+        BEAST_EXPECT(z == Number{});
+        Number y{1'000'000'000'000'001'500, 32000};
+        BEAST_EXPECT((y == Number{1'000'000'000'000'002, 32003}));
+        Number m{std::numeric_limits<std::int64_t>::min()};
+        BEAST_EXPECT((m == Number{-9'223'372'036'854'776, 3}));
```

Comment:
This test fails on Ubuntu 20.04.1 (ARM64). `m` is 0. Should this architecture be supported?

Reply:
Thanks for this report.  Is `m == 0` on the Intel architecture as well?

Reply:
Never mind.  I'm reproducing it now.  This is a bug somewhere in `Number` as it only reproduces with optimizations on.

Reply:
Fixed in https://github.com/XRPLF/rippled/pull/4192/commits/7e1066592a5721aa6f5f5e5bfaee7b8dc73f20bb.

Taking the negative of the most negative signed value is undefined behavior, not identity.  However one can cast negative signed to unsigned and then negate, getting the desired result without UB.

The existing unit tests caught this, and should now pass.